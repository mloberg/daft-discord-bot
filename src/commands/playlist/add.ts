import db from '../../db';
import { FriendlyError } from '../../error';
import logger from '../../logger';
import { hasPermission } from '../../permission';
import player from '../../player';
import { Command } from '../../types';

const command: Command = {
    name: 'playlist add',
    alias: ['playlist append'],
    description: 'Append songs to a playlist',
    usage: '<name> [...songs]',
    async run(message, args) {
        if (!message.member || !hasPermission(message.member)) {
            throw new FriendlyError('You do not have permission to do that.');
        }

        const guild = message.member.guild.id;
        const [name, ...append] = args._;
        if (!name) {
            throw new FriendlyError('I need a name for the playlist.');
        }

        const unsupported = append.filter((s) => !player.supports(s));
        if (unsupported.length !== 0) {
            throw new FriendlyError(`Cannot add to playlist. These songs are unsupported: ${unsupported.join(', ')}`);
        }

        const playlist = await db.playlist.findFirst({ where: { name, guild } });
        if (!playlist) {
            throw new FriendlyError("I couldn't find that playlist.");
        }

        const songs: string[] = playlist.songs as string[];
        songs.push(...append);

        await db.playlist.update({
            where: { id: playlist.id },
            data: { songs },
        });
        logger.info(`Added songs to playlist ${name} in ${message.member.guild.name}`);

        return message.react('ðŸŽµ');
    },
};

export default command;
