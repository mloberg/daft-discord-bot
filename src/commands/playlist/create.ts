import db from '../../db';
import { FriendlyError } from '../../error';
import logger from '../../logger';
import { hasPermission } from '../../permission';
import player from '../../player';
import { Command } from '../../types';

const command: Command = {
    name: 'playlist create',
    alias: ['playlist new'],
    description: 'Create a new playlist',
    usage: '<name> [...songs]',
    async run(message, args) {
        if (!message.member || !hasPermission(message.member)) {
            throw new FriendlyError('You do not have permission to do that.');
        }

        const guild = message.member.guild.id;
        const [name, ...songs] = args._;
        if (!name) {
            throw new FriendlyError('I need a name for the playlist.');
        }

        const unsupported = songs.filter((s) => !player.supports(s));
        if (unsupported.length !== 0) {
            throw new FriendlyError(`Cannot create playlist. These songs are unsupported: ${unsupported.join(', ')}`);
        }

        const existing = await db.playlist.findFirst({ where: { name, guild } });
        if (existing) {
            throw new FriendlyError('There is already a playlist with that name.');
        }

        await db.playlist.create({ data: { name, guild, songs } });
        logger.info(`Created playlist ${name} in ${message.member.guild.name}`);

        return message.react('ðŸŽµ');
    },
};

export default command;
