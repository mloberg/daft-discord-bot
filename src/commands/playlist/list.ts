import db from '../../db';
import { FriendlyError } from '../../error';
import { hasPermission } from '../../permission';
import { Command } from '../../types';

const command: Command = {
    name: 'playlist list',
    alias: ['playlist show'],
    description: 'List all playlists or songs in a playlist',
    usage: '[name]',
    async run(message, args) {
        if (!message.member || !hasPermission(message.member)) {
            throw new FriendlyError('You do not have permission to do that.');
        }

        const guild = message.member.guild.id;
        const name = args._.shift();
        if (!name) {
            const playlists = await db.playlist.findMany({ where: { guild } });

            return message.reply(playlists.map((playlist) => playlist.name).join(', '));
        }

        const playlist = await db.playlist.findFirst({ where: { name, guild } });
        if (!playlist) {
            throw new FriendlyError("I couldn't find that playlist.");
        }

        const songs = playlist.songs as string[];
        return message.reply(songs.join(', '));
    },
};

export default command;
