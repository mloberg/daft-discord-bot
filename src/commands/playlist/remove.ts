import db from '../../db';
import { FriendlyError } from '../../error';
import logger from '../../logger';
import { hasPermission } from '../../permission';
import { Command } from '../../types';

const command: Command = {
    name: 'playlist delete',
    alias: ['playlist remove'],
    description: 'Delete a playlist',
    usage: '<name>',
    async run(message, args) {
        if (!message.member || !hasPermission(message.member)) {
            throw new FriendlyError('You do not have permission to do that.');
        }

        const guild = message.member.guild.id;
        const name = args._.shift();
        if (!name) {
            throw new FriendlyError('Give me a playlist to delete.');
        }

        await db.playlist.deleteMany({ where: { name, guild } });
        logger.info(`Deleted playlist ${name} from ${message.member.guild.name}`);

        return message.react('ðŸŽµ');
    },
};

export default command;
