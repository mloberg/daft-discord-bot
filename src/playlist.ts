import fs from 'fs-extra';
import { difference } from 'lodash';

import { Dictionary } from './types';
import { env } from './utils';

interface Song {
    file: string;
    tags: string[];
}

export class Manager {
    private playlists: Dictionary<string[]> = {};

    constructor(private readonly file: string) {}

    async addSong(file: string, tags: string[]): Promise<void> {
        const songs = await this.getSongs();
        songs.push({ file, tags });
        await this.saveSongs(songs);
    }

    async findSongs(tags: string[]): Promise<string[]> {
        const songs = await this.getSongs();

        return songs.filter((s) => difference(tags, s.tags).length === 0).map((s) => s.file);
    }

    clear(guild: string, room: string): void {
        delete this.playlists[`${guild}_${room}`];
    }

    create(guild: string, room: string, songs: string[]): void {
        this.playlists[`${guild}_${room}`] = songs;
    }

    next(guild: string, room: string): string | null {
        const playlist = this.playlists[`${guild}_${room}`];
        if (!playlist) {
            return null;
        }

        return playlist.shift() || null;
    }

    private async getSongs(): Promise<Song[]> {
        return fs.existsSync(this.file) ? await fs.readJSON(this.file) : [];
    }

    private async saveSongs(songs: Song[]) {
        await fs.writeJSON(this.file, songs);
    }
}

export default new Manager(env('SONGS_FILE', 'songs.json'));
