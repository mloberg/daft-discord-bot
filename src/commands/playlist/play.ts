import { shuffle } from 'lodash';

import db from '../../db';
import { FriendlyError } from '../../error';
import { hasPermission } from '../../permission';
import playlist from '../../playlist';
import { Command } from '../../types';
import next from '../next';

const command: Command = {
    name: 'playlist play',
    alias: ['playlist start'],
    description: 'Play a playlist',
    usage: '<name> [--shuffle|-r] [--volume|-v <volume>]',
    async run(message, args) {
        if (!message.member || !hasPermission(message.member)) {
            throw new FriendlyError('You do not have permission to do that.');
        }

        if (!message.member.voice.channel) {
            throw new FriendlyError('You are not in a voice channel.');
        }

        const guild = message.member.guild.id;
        const room = message.member.voice.channel.name;

        const name = args._.shift();
        if (!name) {
            throw new FriendlyError('Give me a playlist to play.');
        }

        const toPlay = await db.playlist.findFirst({ where: { guild, name } });
        if (!toPlay) {
            throw new FriendlyError("I couldn't find that playlist.");
        }

        const songs = toPlay.songs as string[];
        const randomize = args.shuffle || args.r;

        playlist.clear(guild, room);
        playlist.create(guild, room, randomize ? shuffle(songs) : songs);

        return await next.run(message, args);
    },
};

export default command;
