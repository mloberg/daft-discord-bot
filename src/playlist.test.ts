import fs from 'fs';

import { Manager } from './playlist';

describe('Manager', () => {
    const manager = new Manager('test.json');

    afterEach(() => {
        if (fs.existsSync('test.json')) {
            fs.unlinkSync('test.json');
        }
    });

    it('adds a song', async () => {
        await manager.addSong('test.mp3', ['foo', 'bar']);

        const content = fs.readFileSync('test.json').toString();
        expect(content).toMatchSnapshot();
    });

    it('finds songs matching tags', async () => {
        await manager.addSong('test.mp3', ['foo', 'bar']);
        await manager.addSong('foo.mp3', ['foo']);
        await manager.addSong('bar.mp3', ['bar']);

        expect(await manager.findSongs(['foo'])).toEqual(['test.mp3', 'foo.mp3']);
        expect(await manager.findSongs(['bar'])).toEqual(['test.mp3', 'bar.mp3']);
        expect(await manager.findSongs(['foo', 'bar'])).toEqual(['test.mp3']);
        expect(await manager.findSongs(['testing'])).toEqual([]);
    });

    it('clears a playlist', () => {
        manager.create('foo', 'bar', ['test.mp3']);
        manager.clear('foo', 'bar');
        expect(manager.next('foo', 'bar')).toEqual(null);
    });

    it('creates a playlist', () => {
        manager.create('foo', 'bar', ['test.mp3']);
        expect(manager.next('foo', 'bar')).toEqual('test.mp3');
    });

    it('fetches the next song in the playlist', () => {
        manager.create('foo', 'bar', ['one.mp3', 'two.mp3']);

        expect(manager.next('foo', 'bar')).toEqual('one.mp3');
        expect(manager.next('foo', 'bar')).toEqual('two.mp3');
        expect(manager.next('foo', 'bar')).toEqual(null);
    });
});
