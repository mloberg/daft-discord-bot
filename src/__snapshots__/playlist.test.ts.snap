// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Manager adds a song 1`] = `
"[{\\"file\\":\\"test.mp3\\",\\"tags\\":[\\"foo\\",\\"bar\\"]}]
"
`;
